Implementation of serial passthrough.

btimby opened the pull request:
Meant to be used for accessing the OSD without removing it.

I folded the gps passthrough functionality into this as the two are similar.

I need some review and comments from a more experienced developer. I am not sure I grok the serial system, mostly I have questions about config and how one goes about locating available serial ports etc. Pointers to example code are welcome.

My attempt to fix #1587

btimby committed:
A rough implementation of serial passthrough. Meant to be used for
accessing the OSD without removing it.

I folded the gps passthrough functionality into this as the two
are similar.

btimby commented:
I would like to make it clear that this is not working yet but very close. I need a little input to finalize this.

hydra commented:
what do you need help with?

btimby commented:
I have a couple questions.

1. I would love a code review to let me know if I am on the right track.
2. I am unsure about adding a CLI command, or if it would be better to merge this with gps passthrough.

I am also unsure about the serial system.

1. For the CLI command I chose to use `identifier` to specify the serial port. This is the same as the first value of the serial command. Is this a suitable identifier to expect the user to provide?
2. I am also unsure what parameters are necessary for `openSerialPort` and which ones can be omitted.

Also, I did not see any guidance in the developer documentation on how to actually run the code and verify it works for all targets etc. I assume there is an easier way than loading it onto a board and testing it. I am happy to document this if you kindly point me in the right direction.

btimby committed:
Some cleanup, fixed compiler errors and warnings.

dkisselev commented:
I think replacing GPS passthrough is the solution that makes the most sense, it does basically the exact same thing so this is just a more configurable version of it.

@ledvinap had once implemented serial passthrough as a GPS passthrough replacement, but he never opened a PR to merge it back in, so I'm really glad to see someone take another crack at it. He's one of the main contributors to CF, so looking at his implementation might help you work out any code style or "am I doing it right" questions you may have: https://github.com/ledvinap/cleanflight/blob/local/src/main/io/serial_cli.c#L1609 (linked from https://github.com/cleanflight/cleanflight/issues/1439#issuecomment-151403202).

For testing, a good place to start is the Travis CI results, it looks like the unit test suite fails due to some missing/undefined functions, and your code doesn't seem to fit onto the CJMCU target, so there might be some optimization work needed there.

Once the tests/builds pass, the best next step would probably be to squash your commits and directly ask for people to test your changes with their hardware (bonus points if you compile all the targets and provide easy download links).

Disclaimer, I'm not a contributor on cf, just a passerby that's been following tickets for a while.

dkisselev commented:
Looking at the makefile, CJMCU has half as much flash as any other target, so it makes sense that it would run out early: https://github.com/cleanflight/cleanflight/blob/master/Makefile#L41

It looks like the GPS passthrough function is currently only compiled in when GPS is compiled: https://github.com/cleanflight/cleanflight/blob/master/src/main/io/serial_cli.c#L126 but CJMCU doesn't include GPS because of the small flash size (https://github.com/cleanflight/cleanflight/blob/master/src/main/target/CJMCU/target.h).

A way to solve this might be to add a new compile flag for passthrough, and add it to every target except CJMCU, but that feels like it would be repeating ourselves (since that flag will be true on pretty much every single target in the future).

You could also piggyback on the GPS flag, but that's dirty and will end up being confusing to the next person that looks at the code.

I've seen a few instances of
```
#if (FLASH_SIZE > 64)
#define BLACKBOX
#endif
```

So maybe adding new standard flags such as "128_flash" and "256_flash" that can be used outside of target files could be an option, for features that are "standard" but should only be compiled in when there's enough space, but might lead to an inconsistent experience (although the current setup where we explicitly enable features one-by-one for each target might be no better).

@hydra, I know flash limits are becoming a problem for other features on even the larger targets, do you have any comments on that here or has there been larger discussion about how to handle "deprecated" boards without stopping support entirely?

btimby commented:
Thanks @dkisselev, great info. I could not find the @ledvinap code though I had read mention of "someone" working on this before. Reading his implementation answered most of my questions.

Regarding replacement of GPS passthrough, I agree it may be best from a coding perspective to just implement a "serial pass through" function that supplants the dedicated GPS passthrough. GPS data parsing could be handled by serial passthrough if the port type is GPS or if the user requests it with a CLI flag. However, I do not know how widely used gpspassthrough is and how much resistance there would be to renaming it. My plan is to leave gpspassthrough unchanged, but to share code between it and serial passthrough. Without guidance from @hydra or others, I think this is the best tradeoff. With this in mind I wrote low-level serial passthrough functions that are used by both features.

I think with the information you provided I can complete this code and get a working version.

dkisselev commented:
That's a good point about keeping things backwards compatible.

From your experience with the code, is there any technical reason why we would want GPS passthrough to be implemented any differently from serial passthrough? If possible, the best solution would probably to only keep serial, and strip the gps passthrough command to just be a plain alias that actually runs serial passthrough with any of the necessary parameters.

(ie, instead of building both on the same base, build all the features we could need into serial, and add gps as just a thin wrapper/helper command)

This might be what you're already planning, but the way you explained it sounded like you're planning on keeping more of the GPS stuff around than I'm thinking of.

btimby commented:
The GPS passthrough function does more (GPS specific) stuff than just
shuffle data between serial ports. It takes the data read from the GPS
receiver and passes it to the GPS parsing routines. Additionally, GPS
passthrough puts the LCD display (if present) onto the GPS page and
continues to update it. I am not sure why this is done but my guess is to
keep the CF state updated. Most likely this is so that the user can issue
commands to the GPS unit, see it's reply as well as the effect that has on
CF.

That said, in my implementation GPS passthrough handles the LCD updates,
GPS parsing and uses the low-level serial passthrough functions for data
shuffling.

If you were to combine GPS and serial passthrough you could do one of three
things.

1. Expect the user to request GPS specific handling by using a switch i.e.

serialpassthrough <ID> [BAUD] [TYPE]

Where type could equal GPS or another keyword to invoke special handling.

2. Attempt to detect if the serial port being passed through is GPS. Each
serial port is configured for a use or purpose, GPS is one such. I am not
sure about chances for success here.

3. Both, try to detect but allow the user some control.

btimby committed:
Tested on Naze32 with MinimOSD on UART2, could configure via MWOSD
GUI.

With Cleanflight Configurator open, I saw MSP data from MWOSD GUI.
It was not passed to the OSD. However, once I disconnected CF
Configurator, everything worked.

I will investigate how to detach CF Configurator from the USB port
so that passthrough works without requiring the user to disconnect.

btimby committed:
Ensure the port's mode is full-duplex before relaying.

btimby committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	src/main/io/gps.c

btimby commented:
This works now. I tested serial passthrough to a minimOSD running MWOSD 1.5. I connected it to UART2, which is my usual setup. I did notice that I saw MSP data in the console and MWOSD GUI did not work. It said "Waiting on OSD". I had to disconnect CF configurator to allow successful communication between MWOSD GUI and the minimOSD. I am going to look at an old implementation of 1wire passthrough that used the CLI to see if anything special was done there to cut the CLI out of the communication path. I would prefer if the user need do nothing more than enter the serialpassthrough command to enable passthrough.

I will follow up with Travis to ensure the test suite passes after these changes.

I did **NOT** test GPS passthrough (yet). Also I have not looked into whether it is possible to flash the OSD or not. Reading the related issue leads me to believe it could be done by adding a soft reboot MSP command to MWOSD and also somehow configuring Arduino/AVRDude to send this before flashing. I really have no clue about this.

I also have not tested a bluetooth adapter, but will. I assume I would be able to change it's name or PIN or baud using this method.

btimby committed:
Duh and also add message about uncommitted CLI changes.

hydra commented:
good progress, keep on it!

ledvinap commented on src/main/io/serial.c:
```diff
@@ -421,3 +423,45 @@ void evaluateOtherData(serialPort_t *serialPort, uint8_t receivedChar)
+arbitrary serial passthrough "proxy". Manages LEDs while using the low-level
+`serialPassthroughStep()` to shuffle data both ways.
+*/
+void serialPassthrough(serialPort_t *left, serialPort_t *right)
```
You may define this function as 
```
typedef void serialConsumer(uint8_t);
static void nopConsumer(uint8_t) {}
void serialPassthrough(serialPort_t *left, serialPort_t *right, serialConsumer* leftC, serialConsumer* rightC)
{
     waitForSerialPortToFinishTransmitting(left);
     waitForSerialPortToFinishTransmitting(right);
 
     if(!leftC) 
       leftC = &nopConsumer;
     if(!rightC) 
       rightC = &nopConsumer;

     LED0_OFF;
     LED1_OFF;
 
     while(1) {
        if (serialRxBytesWaiting(left)) {
          LED0_ON;
          uint8_t c = serialRead(left);
          serialWrite(right, c);
          leftC(c);
          LED0_OFF;
        }
         if (serialRxBytesWaiting(right)) {
          LED0_ON;
          uint8_t c = serialRead(right);
          serialWrite(left, c);
          rightC(c);
          LED0_OFF;
        }
     } 
}
```
This way the code should handle generic passthrough and gps.

Also it is probably good idea to light led only when char is copied - human eye is very good at detection extremely low duty cycle, `serialRxBytesWaiting()` overhead may be high enough for led to be slightly glowing.

Also it would be great to implement mechanism to end passthrough without reset - "+++" with guard time was used in (good old) modem days : https://en.wikipedia.org/wiki/Escape_sequence#Modem_control

btimby replied:
Nice suggestion. This cleans up the code a bit. I could not do proper LED management the way I had it.

ledvinap commented on src/main/io/serial_cli.c:
```diff
@@ -970,6 +973,61 @@ static void cliSerial(char *cmdline)
+        }
+
+        passThroughPort = openSerialPort(id, FUNCTION_PASSTHROUGH, NULL,
+                                         baudRates[baud], MODE_RXTX,
```
It is probably better to avoid baudrate table lookup completely or allow user to specify non-default baudrate.
```
if(baud<BAUD_RATE_COUNT)
   baud = baudRates[baud];
```

ledvinap commented on src/main/io/serial_cli.c:
```diff
@@ -970,6 +973,61 @@ static void cliSerial(char *cmdline)
+        printf("Port %d opened, baud=%d\r\n", id, baudRates[baud]);
+    } else {
+        passThroughPort = passThroughPortUsage->serialPort;
+        if (!(passThroughPort->mode & MODE_TX))
```
Forcing RXTX mode may have undesired consequences - what if user wants to test serial receiver - opening port in RXTX mode may claim TX pin that is used for something else (I am not sure current code supports this).

Just noting this limitation is probably OK for now.

ledvinap replied:
Also it would be nice to explicitly call serialSetMode only once (only one `if` will trigger in reality, but the code just looks 'wrong')

btimby replied:
I can easily modify `serialPassThrough()` to "do the right thing" based on what mode the port is opened in by interrogating it's mode. Then it can become a one-way passthrough. Also can support an optional [mode] parameter to override this behavior.

I already fixed the double call to `serialSetMode()` and committed with a **duh** commit message. Just not pushed cleanup changes yet ;-)

ledvinap replied:
`serialRxBytesWaiting()` must return 0/false if port is TX only (or it is buggy).
 `serialWrite()` should handle RX only port without breaking something else.
IMO it is not necessary to handle this explicitly in `serialPassThrough`, the overhead is negligible.

It is safer to keep RX/TX port mode unchanged when port is already open and let user override this.
When opening port, use RXTX by default, but allow override.

Another modes may be interesting in future (inversion/singlewire), but that is usable mostly only for developers

btimby replied:
Got it. No special mode checking necessary.

ledvinap commented:
@btimby : IMO disconnecting configurator is necessary to use serial port in other application, it is strange that MWOSD GUI it did not fail trying to open the port.

After entering pass-through function rest of cleanflight does not run (except ISR routines), so nothing should interfere on this size. This is actually quite bad - `gpspassthrough` has to call `updateDisplay();` explicitly and other problems may appear.
But fixing this may be non-trivial is probably not worth the trouble.

btimby committed:
Modified to allow user to specify the mode and baud rate (as a speed i.e. 115200)

Use callbacks in `serialPassthrough()` to simplify implementation of `gpsPassthrough()`

By default, leave mode untouched for open serial port. If user specifies a mode override it. Do not modify baud for open port.

btimby commented:
@ledvinap Thanks for the input, much appreciated. I tried to address all of your suggestions, I did leave a couple open with TODO comments, I will circle back to these when I have some more time.

Right now I am totally stumped why unit tests fail. The error is:

    clang -g -Wall -pthread -Wextra -ggdb3 -O0 -DUNIT_TEST -isystem ../../lib/test/gtest/inc -MMD -MP -std=gnu99 -Iunit -I../main -c ../main/io/serial.c -o ../../obj/test/io/serial.o
    clang++ -g -Wall -pthread -Wextra -ggdb3 -O0 -DUNIT_TEST -isystem ../../lib/test/gtest/inc -MMD -MP -std=gnu++11 -Iunit -I../main -c unit/io_serial_unittest.cc -o ../../obj/test/io_serial_unittest.o
    clang++ -g -Wall -pthread -Wextra -ggdb3 -O0 -DUNIT_TEST -isystem ../../lib/test/gtest/inc -MMD -MP -std=gnu++11 ../../obj/test/io/serial.o ../../obj/test/io_serial_unittest.o ../../obj/test/gtest_main.a -o ../../obj/test/../../obj/test/io_serial_unittest
    ../../obj/test/io/serial.o: In function `serialPassthrough':
    /home/travis/build/cleanflight/cleanflight/src/test/../main/io/serial.c:460: undefined reference to `serialRxBytesWaiting'
    /home/travis/build/cleanflight/cleanflight/src/test/../main/io/serial.c:462: undefined reference to `serialRead'
    /home/travis/build/cleanflight/cleanflight/src/test/../main/io/serial.c:463: undefined reference to `serialWrite'
    /home/travis/build/cleanflight/cleanflight/src/test/../main/io/serial.c:467: undefined reference to `serialRxBytesWaiting'
    /home/travis/build/cleanflight/cleanflight/src/test/../main/io/serial.c:469: undefined reference to `serialRead'
    /home/travis/build/cleanflight/cleanflight/src/test/../main/io/serial.c:470: undefined reference to `serialWrite'
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    make: *** [../../obj/test/io_serial_unittest] Error 1

ledvinap commented:
This is quite common with tests.
Juste define empty stub functions here: https://github.com/cleanflight/cleanflight/blob/master/src/test/unit/io_serial_unittest.cc#L64

ledvinap commented on src/main/io/gps.c:
```diff
@@ -1038,6 +1038,16 @@ static bool gpsNewFrameUBLOX(uint8_t data)
     return parsed;
 }
 
+void gpsHandlePassthrough(uint8_t data)
```
declare this function `static`, there is no need to inflate global namespace

ledvinap commented on src/main/io/serial_cli.c:
```diff
@@ -970,6 +974,82 @@ static void cliSerial(char *cmdline)
+                break;
+            case 1:
+                baud = atoi(tok);
+                baudIdx = lookupBaudRateIndex(baud);
```
Is it really necessary to force one of enumerated baudrates? Let user shoot his leg if he insists ...

ledvinap commented on src/main/io/serial_cli.c:
```diff
@@ -970,6 +974,82 @@ static void cliSerial(char *cmdline)
+            mode = MODE_RXTX;
+
+        passThroughPort = openSerialPort(id, FUNCTION_PASSTHROUGH, NULL,
+                                         baudIdx, MODE_RXTX,
```
`mode` should be used here ...

ledvinap replied:
and actual baudrate is passed, not index

ledvinap commented on src/main/io/serial_cli.c:
```diff
@@ -970,6 +974,82 @@ static void cliSerial(char *cmdline)
+        // If the user supplied a mode, override the port's mode, otherwise
+        // leave the mode unchanged. serialPassthrough() handles one-way ports.
+        printf("Port %d already open\r\n", id);
+        if (mode && !(passThroughPort->mode & mode)) {
```
This won't work as intended ( consider `MODE_TX & MODE_RXTX` ).

Either allow changing mode completely `(passThroughPort->mode != mode)`  (and use `mode` in serialSetMode)
or allow enabling other direction `(mode & ~passThroughPort->mode)`

ledvinap commented on src/main/io/serial_cli.c:
```diff
@@ -970,6 +974,82 @@ static void cliSerial(char *cmdline)
+    }
+
+    printf("********************************************************************************\r\n");
+    printf("* Relaying data to device on port %d, You must disconnect or close Cleanflight *\r\n", id);
```
The 'close port to open it in other application' part may be moved to documentation.. 

It is possible to use connected device in Configurator directly (if device is text based) or enter CLI in other terminal program (just send '#' character)

ledvinap commented:
Great work!

ledvinap commented on src/main/io/gps.c:
```diff
@@ -1054,27 +1064,7 @@ void gpsEnablePassthrough(serialPort_t *gpsPassthroughPort)
-        }
-#endif
-    }
+    serialPassthrough(gpsPort, gpsPassthroughPort, &gpsHandlePassthrough, NULL);
```
seriaPassthrough handles leds internally, no need to turn them off in this function

btimby committed:
Define stubs to allow serial unit tests to execute.

btimby committed:
Define gpsHandlePassthrough as static it is not a globally useful function.

btimby committed:
Allow user to define baud rate exactly as desired.
Fix passing mode and baud rate to openSerialPort().
Fix mode handling so that mode is set as user specified.

btimby committed:
Add serialpassthrough command documentation.

btimby committed:
Ammend Serial documentation to mention passthrough mode.

btimby committed:
Oops found while testing.

btimby commented:
I believe I have addressed all feedback, the unit tests pass and I have updated all documentation.

I tested this with an OSD connected. I tried with the port configured and unconfigured i.e.

    serialpassthrough 1

To connect to port 1 (UART 2) as configured. and...

    serialpassthrough 1 115200

To connect to port 1 (UART 2) with the baude rate specified.

I have not yet tested GPS passthrough, I will need to dig one up and configure it (I don't use GPS).

btimby commented:
I completed this feature to allow me to upload a font to my micro minimOSD which is buried and direct soldered to my Naze32 inside an RX122. This way I don't have to tear the copter apart to address this problem (I forgot this step during my build). This is also useful to reconfigure the OSD (esp. change NTSC/PAL) or modify layouts.

It is not currently possible to flash firmware to the OSD as it requires a reset. I don't know if there is a software workaround for this or not.

ledvinap commented on docs/Serial.md:
```diff
@@ -81,3 +81,22 @@ The allowable baud rates are as follows:
+
+    serialpassthrough <id> [baud] [mode]
+
+ID is the index of the serial port to pass through communication to. This index is 0-based and matches the ports on the Cleanflight Configurator ports page (see above). Baud is the desired baud rate, and mode is a combination of the keywords rx and tx (rxtx is full duplex). The baud and mode parameters can be used to override the configured values for the specified port.
```
https://github.com/cleanflight/cleanflight/blob/f17ef30bb8b9e0156b6bb72b766d4e5acc1b4add/src/main/io/serial.h#L54-L60

ledvinap commented:
these led command may be removed: https://github.com/cleanflight/cleanflight/pull/1620/files#diff-62a123af09bd3b80f478b220998900f3R1059

btimby committed:
Remove redundant LED control code.

btimby committed:
Updated documentation for correctness.

btimby committed:
Remove source code link and replace with reference to serialPortIdentifier_e).

btimby committed:
Remove some whitespace in stub functions.

btimby committed:
Remove non-standard strcasestr() in favor of standard strstr(). Hoping this fixes warning in unittest compilation.

btimby committed:
Make serial passthrough configurable, disable for CJMCU target.

btimby committed:
Merge remote-tracking branch 'upstream/master'

Conflicts:
	src/main/target/CJMCU/target.h

btimby committed:
If skipping serialpassthrough when GPS is enabled, we need these functions for gpsPassthrough to function.

btimby commented:
@hydra I think this is ready for prime time. I was able to test GPS passthrough using the following method. I don't use GPS, so this is what I was able to come up with:

1. I connected a ublox NEO-6 GPS to UART2 of my Naze32
2. I ran gpspassthrough in the CF CLI
3. I was able to run `gpsd -N -S 10004 -D 2 /dev/ttyUSB0` 
4. I could connect to GPSd via telnet and query it for GPS data

```bash
    $ telnet localhost 10004
    Trying 127.0.0.1...
    Connected to localhost.
    Escape character is '^]'.
    {"class":"VERSION","release":"3.9","rev":"3.9","proto_major":3,"proto_minor":8}
    ?WATCH={"enable":true}
    {"class":"DEVICES","devices":[{"class":"DEVICE","path":"/dev/ttyUSB0","activated":"2016-01-15T18:23:34.608Z","native":0,"bps":115200,"parity":"N","stopbits":1,"cycle":1.00}]}
    {"class":"WATCH","enable":true,"json":false,"nmea":false,"raw":0,"scaled":false,"timing":false}
    {"class":"DEVICE","path":"/dev/ttyUSB0","activated":"2016-01-15T18:23:34.618Z","driver":"uBlox UBX binary","native":1,"bps":115200,"parity":"N","stopbits":1,"cycle":1.00,"mincycle":0.25}
    ?POLL;
    {"class":"POLL","time":"2016-01-15T18:23:59.364Z","active":1,"tpv":[{"class":"TPV","tag":"0x0120","device":"/dev/ttyUSB0","mode":1,"time":"2016-01-15T18:24:01.200Z","ept":0.005}],"gst":[{"class":"GST","tag":"0x0120","device":"/dev/ttyUSB0","time":"1970-01-01T00:00:00.000Z","rms":0.000,"major":0.000,"minor":0.000,"orient":0.000,"lat":0.000,"lon":0.000,"alt":0.000}],"sky":[{"class":"SKY","tag":"0x0120","device":"/dev/ttyUSB0","vdop":99.99,"tdop":99.99,"hdop":99.99,"gdop":99.99,"pdop":99.99,"satellites":[{"PRN":10,"el":-91,"az":0,"ss":39,"used":false}]}]}
```

To me this indicates that the Naze32 is passing bidirectional serial data to the GPS device and that GPSd could communicate with it.

nathantsoi commented:
@btimby, first let me say. awesome work! would you have the bandwidth to port this to https://github.com/borisbstyle/betaflight I'd love to see this get merged in (to betaflight at least)!

btimby commented:
@nathantsoi Thanks! I will see what I can do regarding betaflight.

btimby committed:
Disable buzzer during passthrough.

I had the buzzer remain beeping during my entire passthrough session the other day.

btimby commented:
@nathantsoi I have created a PR for betaflight. I was not previously aware that I could target forks when creating a PR. Easy peasy.

nathantsoi commented:
nice work @btimby! sounds like ppl love this :) i need a fc with a vcp to test it out :)

btimby committed:
Add MSP command to enable serial passthrough. This makes a few desirable things possible.

 - Allow the cleanflight GUI to initiate passthrough (like a button on ports config tab).
 - Allow the MWOSD GUI to initiate passthrough.

Also, I know Dominic is keen to remove the CLI at some point and supporting this in MSP would be required.

I lifted the idea from this MultiWii proposal:

http://www.multiwii.com/forum/viewtopic.php?p=62207

btimby committed:
Added escape sequence to exit serial passthrough mode.

drkavnger99 commented:
Is there any updates on when this may be rolled into cleanflight?  I'm using it on betaflight and it works really well.

fhuber commented:
This is such a useful feature. I hope it will make it into the next release.

hydra commented:
@btimby gonna merge this for the v1.14.0 release. it's quite useful indeed.

btimby commented:
@hydra great news, thanks!

Edson2889 commented:
Which pind I have use for pwm on the beerotor f3
